src/interface/swarm.cpp

```diff
@@ -78,6 +87,73 @@ Swarm::Swarm(const std::string &label, const Metadata &metadata, const int nmax_
   marked_for_removal_.data.DeepCopy(marked_for_removal_h);
 }
 
+void Swarm::AllocateBoundaries() {
+  auto pmb = GetBlockPointer();
+  std::stringstream msg;
+
+  auto &bcs = pmb->pmy_mesh->mesh_bcs;
+
+  if (bcs[0] == BoundaryFlag::outflow) {
+    bounds[0] = DeviceAllocate<ParticleBoundIX1Outflow>();
+  } else if (bcs[0] == BoundaryFlag::periodic) {
+    bounds[0] = DeviceAllocate<ParticleBoundIX1Periodic>();
+  } else if (bcs[0] != BoundaryFlag::user) {
+    msg << "ix1 boundary flag " << static_cast<int>(bcs[0]) << " not supported!";
+    PARTHENON_THROW(msg);
+  }
+
+  if (bcs[1] == BoundaryFlag::outflow) {
+    bounds[1] = DeviceAllocate<ParticleBoundOX1Outflow>();
+  } else if (bcs[1] == BoundaryFlag::periodic) {
+    bounds[1] = DeviceAllocate<ParticleBoundOX1Periodic>();
+  } else if (bcs[1] != BoundaryFlag::user) {
+    msg << "ox1 boundary flag " << static_cast<int>(bcs[1]) << " not supported!";
+    PARTHENON_THROW(msg);
+  }
+
+  if (bcs[2] == BoundaryFlag::outflow) {
+    bounds[2] = DeviceAllocate<ParticleBoundIX2Outflow>();
+  } else if (bcs[2] == BoundaryFlag::periodic) {
+    bounds[2] = DeviceAllocate<ParticleBoundIX2Periodic>();
+  } else if (bcs[2] != BoundaryFlag::user) {
+    msg << "ix2 boundary flag " << static_cast<int>(bcs[2]) << " not supported!";
+    PARTHENON_THROW(msg);
+  }
+
+  if (bcs[3] == BoundaryFlag::outflow) {
+    bounds[3] = DeviceAllocate<ParticleBoundOX2Outflow>();
+  } else if (bcs[3] == BoundaryFlag::periodic) {
+    bounds[3] = DeviceAllocate<ParticleBoundOX2Periodic>();
+  } else if (bcs[3] != BoundaryFlag::user) {
+    msg << "ox2 boundary flag " << static_cast<int>(bcs[3]) << " not supported!";
+    PARTHENON_THROW(msg);
+  }
+
+  if (bcs[4] == BoundaryFlag::outflow) {
+    bounds[4] = DeviceAllocate<ParticleBoundIX3Outflow>();
+  } else if (bcs[4] == BoundaryFlag::periodic) {
+    bounds[4] = DeviceAllocate<ParticleBoundIX3Periodic>();
+  } else if (bcs[4] != BoundaryFlag::user) {
+    msg << "ix3 boundary flag " << static_cast<int>(bcs[4]) << " not supported!";
+    PARTHENON_THROW(msg);
+  }
+
+  if (bcs[5] == BoundaryFlag::outflow) {
+    bounds[5] = DeviceAllocate<ParticleBoundOX3Outflow>();
+  } else if (bcs[5] == BoundaryFlag::periodic) {
+    bounds[5] = DeviceAllocate<ParticleBoundOX3Periodic>();
+  } else if (bcs[5] != BoundaryFlag::user) {
+    msg << "ox3 boundary flag " << static_cast<int>(bcs[5]) << " not supported!";
+    PARTHENON_THROW(msg);
+  }
+
+  for (int n = 0; n < 6; n++) {
+    pbounds.bounds[n] = bounds[n].get();
+    PARTHENON_REQUIRE(pbounds.bounds[n] != nullptr,
+                      "Null device boundary condition pointer!");
+  }
```

Comment:
I think this error message could be more useful for the confused downstream user. Something like:
```suggestion
  for (int n = 0; n < 6; n++) {
    pbounds.bounds[n] = bounds[n].get();
    std::stringstream msg;
    msg << "Boundary condition on face " << n << " missing.\n"
           << "Please set it to `outflow`, `periodic`, or `user` in the input deck.\n"
           << "If you set it to user, you must also manually set "
           << "the swarm boundary pointer in your application."
           << std::endl;
    PARTHENON_REQUIRE(pbounds.bounds[n] != nullptr, msg);
  }
```